src/software/util/design_patterns/constructor_arg_generic_factory.h

```diff
@@ -0,0 +1,187 @@
+#pragma once
+#include <any>
+#include <functional>
+#include <memory>
+#include <string>
+#include <sstream>
+#include <unordered_map>
+#include <vector>
+#include "software/util/typename/typename.h"
+//#include "software/backend/backend.h" // Should make generic.h
+
+// A quality of life typedef to make things shorter and more readable
+template <class IndexType, class TypeToCreate>
+using ConstructorArgGenericRegistry =
+std::unordered_map<IndexType, std::function<std::unique_ptr<TypeToCreate>(std::any)>>;
+/**
+ * The GenericFactory is an Abstract class that provides an interface for Generic type
+ * Factories to follow. This makes it easy to maintain a list of factories and get the
+ * corresponding generic types through the generic interface.
+ */
+
+// had to roll our own function_traits class to get function arg types as a tuple
+template <typename Function> class fn_traits;
+
+template <typename ResultType, typename... ArgTypes>
+class fn_traits<ResultType(ArgTypes...)>
+{
+   public:
+    using args_tuple_type = std::tuple<ArgTypes...>;
+    using result_type = ResultType;
+};
+
+
+// modified from std::make_from_tuple in libstdc++ with names de-uglified
```

Comment:
It wasn't just names that were changed, the function itself was also changed to construct a `unique_ptr<CreatedType>` instead of a `CreatedType`. This couldn't be done with `std::make_from_tuple`.

Reply:
Did this get moved? I don't see it now.